Frameworks/QuartzCore/CALayer.mm

```diff
@@ -353,19 +366,31 @@ void ReleaseDisplayTexture(DisplayTexture* tex) {
 static LockingBufferInterface _globallockingBufferInterface;
 
 CGContextRef CreateLayerContentsBitmapContext32(int width, int height, float scale) {
-    DisplayTexture* tex = NULL;
-
     if ([NSThread isMainThread]) {
-        tex = GetCACompositor()->CreateWritableBitmapTexture32(width, height);
-    }
+        DisplayTexture* writetableBitmapTexture = GetCACompositor()->CreateWritableBitmapTexture32(width, height);
+        RETURN_NULL_IF(!writetableBitmapTexture);
+        auto releaseWritetableBitmapTexture = wil::ScopeExit([&]() { GetCACompositor()->ReleaseDisplayTexture(writetableBitmapTexture); });
+
+        std::unique_ptr<IDisplayTextureImpl> displayTexture(new IDisplayTextureImpl(writetableBitmapTexture));
+        releaseWritetableBitmapTexture.Dismiss();
 
-    CGContextRef ret = _CGBitmapContextCreateWithTexture(width, height, scale, tex, &_globallockingBufferInterface);
+        Microsoft::WRL::ComPtr<IWICBitmap> customWICBtmap =
+            Microsoft::WRL::Make<CGIWICBitmap>(displayTexture.get(), GUID_WICPixelFormat32bppPBGRA, height, width);
+        // We want to convert it to GUID_WICPixelFormat32bppPBGRA for D2D Render Target compatibility.
+        woc::unique_cf<CGImageRef> image(_CGImageCreateWithWICBitmap(customWICBtmap.Get()));
 
-    if (tex) {
-        _globallockingBufferInterface.ReleaseDisplayTexture(tex);
+        Microsoft::WRL::ComPtr<ID2D1Factory> factory;
+        RETURN_NULL_IF_FAILED(_CGGetD2DFactory(&factory));
+
+        Microsoft::WRL::ComPtr<ID2D1RenderTarget> renderTarget;
+        RETURN_NULL_IF_FAILED(factory->CreateWicBitmapRenderTarget(customWICBtmap.Get(), D2D1::RenderTargetProperties(), &renderTarget));
+        renderTarget->SetDpi(kWindowsDPI * scale, kWindowsDPI * scale);
```

Comment:
we can also pull c_windowsDPI down here where it's used since it's not used anywhere else.

---
In reply to: [87295048](https://github.com/Microsoft/WinObjC/pull/1332#discussion_r87295048) [](ancestors = 87295048,87294346)